### ISSUE TITLE
Getting confusing error

### ISSUE THREAD

**Issue by jonlaing** [davidkpiano/xstate#506]
## Bug or feature request?

More bug, but probably just need some help figuring this out.

## Description:

Was doing some refactoring and upgraded xstate to 4.6.2 and now I'm getting this crazy error on React Native. 

```
actions_1.invoke is not a function. (In 'actions_1.invoke(invokeConfig)', 'actions_1.invoke' is undefined)
```

The stack trace points me to `StateNode.js:139:36`, but nothing is jumping out at me there as to what the issue might be. I'm pretty sure the Machine it's complaining about is the following:

```typescript
import {
  MachineConfig,
  ActionFunctionMap,
  Machine,
  ServiceConfig,
  ConditionPredicate,
  EventObject,
  InvokeCreator
} from "xstate";
import { SalesorderlineShape } from "../../templates/salesorderline";
import { assign, log } from "xstate/lib/actions";
import { hasLocal } from "../FetchAndMerge";
import { initCart } from "./actions";
import refreshCart from "../CommonActions/refreshCart";
import { SalesorderShape } from "../../templates/salesorder";
import { PosProfileShape } from "../../templates/posProfile";
import { StoreShape } from "../../templates/store";

export interface CartInitSchema {
  states: {
    checkContextCart: {};
    checkLocal: {};
    initCart: {};
    refreshCart: {};
    initCartFail: {};
    refreshCartFail: {};
    done: {};
  };
}

export interface CartInitContext {
  // determines whether you can look at a closed cart from here.
  // If false and the cart in context is closed, it will init
  // a new cart
  allowClosed: boolean;
  posProfile: PosProfileShape;
  shoppingCart: SalesorderShape;
  items: SalesorderlineShape[];
  cache: boolean;
  error?: any;
}

export const initialContext: CartInitContext = {
  allowClosed: true,
  posProfile: null,
  shoppingCart: null,
  items: [],
  cache: true
};

export const Chart: MachineConfig<CartInitContext, CartInitSchema, any> = {
  id: "cart",
  initial: "checkContextCart",
  context: initialContext,
  states: {
    checkContextCart: {
      on: {
        "": [
          { target: "refreshCart", cond: "hasContextCart" },
          { target: "checkLocal" }
        ]
      }
    },
    checkLocal: {
      invoke: {
        src: "checkLocalCart",
        onDone: [
          {
            target: "refreshCart",
            actions: "updateCart",
            cond: "isCartOpen"
          },
          { target: "initCart" }
        ],
        onError: "initCart"
      }
    },
    initCart: {
      invoke: {
        src: "initCart",
        onDone: {
          target: "fetchCartItems",
          actions: "updateCart"
        },
        onError: "initCartFail"
      }
    },
    refreshCart: {
      invoke: {
        src: "refreshCart",
        onDone: {
          target: "fetchCartItems",
          actions: ["updateCart", log((ctx, evt) => evt, "REFRESHED CART")]
        },
        onError: "refreshCartFail"
      }
    },
    initCartFail: {
      onEntry: log((ctx, evt) => evt, "INIT FAIL"),
      type: "final",
      data: {
        error: (ctx, evt) => evt.data
      }
    },
    refreshCartFail: {
      onEntry: log((ctx, evt) => evt, "REFRESH FAIL"),
      type: "final",
      data: {
        error: (ctx, evt) => evt.data
      }
    },
    done: {
      type: "final",
      data: {
        shoppingCart: ctx => ctx.shoppingCart
      }
    }
  }
};

export const actions: ActionFunctionMap<CartInitContext, EventObject> = {
  updateCart: assign({ shoppingCart: (ctx, evt) => evt.data })
};

export const services: Record<string, ServiceConfig<CartInitContext>> = {
  checkLocalCart: (ctx: CartInitContext) =>
    hasLocal("shoppingCart", (cart: SalesorderShape) => {
      if (typeof ctx.posProfile.channel === "string") {
        return cart.channel === ctx.posProfile.channel;
      }
      return cart.channel === (ctx.posProfile.channel as StoreShape).$uri;
    }),
  initCart: ((ctx, evt) =>
    initCart({
      posProfileURI: ctx.posProfile.$uri
    })) as InvokeCreator<any, any>,
  refreshCart: (ctx =>
    refreshCart(ctx.shoppingCart.$uri, ctx.cache)) as InvokeCreator<
    CartInitContext
  >
};

export const guards: Record<
  string,
  ConditionPredicate<CartInitContext, any>
> = {
  hasContextCart: ctx => {
    return ctx.allowClosed == null || ctx.allowClosed
      ? !!ctx.shoppingCart
      : ctx.shoppingCart && ctx.shoppingCart.status === "Open";
  },
  isCartOpen: (ctx, evt: EventObject & { data: SalesorderShape }) =>
    evt.data.status === "Open"
};

export default Machine(Chart, { actions, services, guards });
```

This was working before the upgrade, and I don't think I changed this particular file in my refactor. I'm wondering if something about invoking Promises changed between version 4.3.2 and version 4.6.2. Any help would be greatly appreciated. Also, I blew away my node_modules once and reinstalled hoping that would solve the problem, but I'm still stuck here. Nothing is jumping out at me as obviously wrong, but if someone could even point me in the right direction that would be awesome. Thanks!

---------

**EDIT:** I'm pretty sure this was a code caching problem on my end. It went away when I forced the cache to clear. I thought I had cleared it before, but apparently not.

**Comment by davidkpiano** [comment]
Can you create a [CodeSandbox (XState template)](https://codesandbox.io/s/xstate-example-template-qcwed) reproduction of this?

And I don't see where `actions_1.invoke(invokeConfig)` is being called, maybe you left that call out? That would be the issue.

**Comment by criso** [comment]
This may be related:
 I’m running into a an issue with spawning a service when using 4.6.2. It works fine on 4.6.0.
The spawned service’s state is missing data.
I’ve been able to replicate the bug against the `todo` example when using 4.6.2

https://codesandbox.io/s/xstate-todomvc-nxxmd

**Comment by jonlaing** [comment]
Okay, well, my error went away. No idea what the hell that was. I was getting a new one though, and I think I've tracked this one down to the source.

I was getting an issue when a promise rejected, and the error I got was related to React Native's own polyfill of `Object.assign`. It fails on `hasOwnProperty`. I'm not sure if this is still an issue in the latest React Native, but I'm using expo so I'm a couple versions behind.

The offending code in RN is here: https://github.com/expo/react-native/blob/exp-latest/Libraries/polyfills/Object.es6.js#L54

The error I get is: "One of the sources for assign has an enumerable key on the prototype chain. Are you trying to assign a prototype property? We don't allow it, as this is an edge case that we do not support. This error is a performance optimization and not spec compliant."

For the record, the key it's breaking on is `inert` on a state object. Not sure what that's used for, but that's the one it doesn't like.

I think this is related to the new spawning behavior in 4.6.x. I downgraded to 4.5.0, as I wasn't using the actor model stuff anyway and that fixed the issue for me. I don't really expect this library to address the weirdness in React Native, but I wanted to leave this comment here for posterity. I'm not even sure if it could be fixed from this end without breaking something else.

Maybe we can put it in a troubleshooting doc for any other poor souls that run into this?
